nixos/modules/services/networking/tinydns.nix

```diff
@@ -1,55 +1,279 @@
+              sent to secondaries via SSH.
+            '';
+          };
+
+          secondary.enable = mkEnableOption "Configure this tinydns instance as a secondary server receiving its data via SSH.";
+
+          secondary.sshKey = mkOption {
+            type = types.str;
+            description = "SSH key to authenticate zone transfers to the secondary DNS server.";
+            default = "";
+          };
+        };
+      }));
     };
   };
 
   ###### implementation
 
-  config = mkIf config.services.tinydns.enable {
+  config = mkIf (any id (mapAttrsToList (_: v: v.enable) config.services.tinydns)) {
+
     environment.systemPackages = [ pkgs.djbdns ];
 
-    users.users.tinydns.isSystemUser = true;
+    # Restrict to a single session at a time
+    services.openssh.extraConfig = ''
+      Match User tinydns
+      MaxSessions=1
+      Match All
+    '';
+
+    users.users.tinydns = {
+      isSystemUser = true;
+      uid = config.ids.uids.tinydns;
+      group = "tinydns";
+      shell = "/run/current-system/sw/bin/bash";
+      openssh.authorizedKeys.keys = mapAttrsToList (name: cfg: let
+        # Expects tinydns data on standard input
+        tinydnsReceive = pkgs.writeScript "tinydns-receive" ''
+        #!${pkgs.stdenv.shell} -e
+        cd ${cfg.rootPath}
+        (
+          flock -n 5 || exit 1
+          mv data data.backup
+          cat >data
+          if ${getBin pkgs.djbdns}/bin/tinydns-data ; then
+            rm data.backup
+          else
+            mv data.backup data
+          fi
+          # Do not allow zone transfers to run more often than once per second due to
+          # minimum DNS record time resolution.
+          sleep 1
+        ) 5>data-lock
+        '';
+      in optionalString (cfg.secondary.sshKey != "") ''
+        restrict,command="${tinydnsReceive}" ${cfg.secondary.sshKey}
+      '') config.services.tinydns;
+    };
+    users.groups.tinydns = {
+      gid = config.ids.gids.tinydns;
+    };
 
-    systemd.services.tinydns = {
+    systemd.services = (mapAttrs' (ip: cfg: nameValuePair "tinydns-${ip}" {
+      enable = cfg.enable;
       description = "djbdns tinydns server";
       wantedBy = [ "multi-user.target" ];
-      after = [ "network.target" ];
-      path = with pkgs; [ daemontools djbdns ];
+      after = [ "network-online.target" ];
+      requires = [ "network-online.target" ];
+      environment = {
+        IP = ip;
+        ROOT = cfg.rootPath;
+        UID = "${toString config.users.users.tinydns.uid}";
+        GID = "${toString config.users.groups.tinydns.gid}";
+        AXFR_JOBS = "2"; # Number of concurrent zone transfers to secondary servers.
+      };
+      path = with pkgs; [ coreutils djbdns ];
+      serviceConfig = {
+        # Drops privileges and chroots to ROOT by itself
+        ExecStart = "${getBin pkgs.djbdns}/bin/tinydns";
+        LimitDATA = mkDefault "40000000";
+        RestartSec = 1;
```

Comment:
I guess this is because of the one second sleep in the script of the receiving end of the zone transfer, so please add a comment here to make sure this is clear.

Reply:
No it's not. The default value was just needlessly slow and there's no point in losing 1000 requests just because your server sleeps for 5 seconds (I believe that's the default).